### PR TITLE
Add documentation, small refactorings

### PR DESCRIPTION
**Pull request by quchen**
I tried documenting a couple of low-hanging fruits here.

One thing I could not wrap my head around is documentation of `WayUnit`s fields, I'd appreciate if someone more knowledgeable could have a look at that. `Way` confuses me!

### CODE REVIEW

**Comment by snowleopard** [comment]
This looks great, many thanks. One thing I'd like to do a bit differently is `src/Main.hs`: can we use fully qualified names for imports? I think `Rules.Cabal.cabalRules` would look better than `RCabal.cabalRules`.

I'll do a full revision a bit later.

**Comment by quchen** [comment]
Fully qualified names were a bit long when I tried them, but if you like it we can do that as well. The main point here was getting rid of the comments, because they described something easily expressible by code (and checked by the compiler).

**Comment by snowleopard** [comment]
> The main point here was getting rid of the comments, because they described something easily expressible by code (and checked by the compiler).

Yes, absolutely, that's definitely a win on all accounts. I think fully qualified names make this even better as one doesn't need to look up the imports.

File: src/Predicates.hs [hunk 1]
```diff
@@ -7,49 +7,60 @@ module Predicates (
...
 notStage0 :: Predicate
 notStage0 = notM stage0
 
+-- | Is a certain package /not/ built right now?
 notPackage :: Package -> Predicate
 notPackage = notM . package
 
--- TODO: Actually, we don't register compiler in some circumstances -- fix.
+-- | TODO: Actually, we don't register compiler in some circumstances -- fix.
 registerPackage :: Predicate
```
Comment (snowleopard) [inline src/Predicates.hs]: TODO should probably be outside haddock. The haddock comment can be:
-- | Do we need to run `ghc-pkg update` on the currently built package? See `Rules.Data.buildPackageData`

File: src/Target.hs [hunk 2]
```diff
@@ -12,50 +16,53 @@ import Package
...
+--
+-- * the empty computation is the identity element of the underlying type
 -- * two computations can be combined by combining their results
 instance Monoid a => Monoid (ReaderT Target Action a) where
     mempty  = return mempty
     mappend = liftM2 mappend
 
--- PartialTarget is a partially constructed Target with fields Stage and
--- Package only. PartialTarget's are used for generating build rules.
+-- A partially constructed Target with fields 'Stage' and 'Package' only.
```
Comment (snowleopard) [inline src/Target.hs]: Should be "-- | A partially..." (missing '|')?

File: src/Way.hs [hunk 3]
```diff
@@ -72,11 +80,23 @@ instance Read Way where
 instance Eq Way where
     Way a == Way b = a == b
 
-vanilla, profiling, logging, parallel, granSim :: Way
+-- | Build with no 'WayUnit's at all.
+vanilla :: Way
 vanilla   = wayFromUnits []
```
Comment (snowleopard) [inline src/Way.hs]: Why extra spaces? Also see logging, parallel, granSim.

**Comment by snowleopard** [comment]
I've added some comments. I think your description of `WayUnit` is pretty close to how I understand it, although I have to admit I'm also a bit hazy on this topic. It's likely ways will be reworked in future.

Many thanks again!

**Comment by quchen** [comment]
Just FYI, you can merge this whenever the build passes and you're happy with the changes.

I'll keep adding commits in the same spirit and rebase it onto your master periodically.

**Comment by snowleopard** [comment]
There are some merge conflicts at the moment.

**Comment by snowleopard** [comment]
I think the conflicts are because of this commit: c573af0.

@ndmitchell Could you please use pull requests, so I can keep track of what is happening? Otherwise you seem to have pushed exactly what @quchen has proposed in this PR :)

**Comment by ndmitchell** [comment]
Opps, sorry for that! Yep, I'll go through pull requests. Perhaps remove my commit permissions? It's very easy to accidentally send to the wrong branch, and far easier to have GitHub stop me.

**Comment by ndmitchell** [comment]
One easy option might be to revert my patch (just `git revert c573af0`), and then this branch should merge cleanly once more.

**Comment by snowleopard** [comment]
@ndmitchell No problem! Yes, I'll do the revert. 

How do I remove your commit permissions without removing from collaborators? I'd like to keep you in collaborators because I can then assign issues to you :-)

**Comment by snowleopard** [comment]
This is now merged.

**Comment by ndmitchell** [comment]
Don't know how assigning really works - I don't think I've ever done it. I will just be careful then.

**Pull request merged**